sympy/concrete/tests/test_sums_products.py

```diff
@@ -1143,12 +1143,21 @@ def test_Sum_dummy_eq():
     assert not Sum(x, (x, a, b)).dummy_eq(Sum(x, (x, a, b), (a, 1, 2)))
     assert not Sum(x, (x, a, b)).dummy_eq(Sum(x, (x, a, c)))
     assert Sum(x, (x, a, b)).dummy_eq(Sum(x, (x, a, b)))
+    assert Sum(x, (x, a, c)).dummy_eq(Sum(y, (y, a, c)))
+    assert Sum(x, (x, a, c)).dummy_eq(Sum(y, (y, a, c)), map=True) == {}
     d = Dummy()
     assert Sum(x, (x, a, d)).dummy_eq(Sum(x, (x, a, c)), c)
-    assert not Sum(x, (x, a, d)).dummy_eq(Sum(x, (x, a, c)))
-    assert Sum(x, (x, a, c)).dummy_eq(Sum(y, (y, a, c)))
+    assert Sum(x, (x, a, d)).dummy_eq(Sum(x, (x, a, c)))
```

Comment:
This doesn't seem right. Surely only x is a dummy variable here. If c and d are unequal then these two sums are unequal.

Reply:
There are two types of dummy symbols: structurally bound (e.g. the summation variable) and physical Dummy symbols. `Expr.as_dummy` will make two expressions that differ only in bound symbols the same. `Expr.dummy_eq` is more of a special form of matching and answers the question of whether two expressions are the same except for names of symbols. And by default it limits attention to the Dummy symbols of self.

In this case, d is the Dummy and the two expression are the same if `d -> c` so that's why `dummy_eq` reports True.

Often this routine can be used in testing. In your PR I noticed that `trigintegrate(cos(x)*cot(x), x)` gives something equivalent to
```python
>>> eq
(-_b1 + _b2)*log(cos(x) + 1)/2 + (_b1 + _b2)*log(cos(x) - 1)/2 + cos(x) + 1
```
where the symbols with leading underscores are Dummy symbols. To test this in this PR you can do
```python
_b1, _b2 = Dummy(), Dummy()  # note that these are now different than those in eq
ans = (-_b1 + _b2)*log(cos(x) + 1)/2 + (_b1 + _b2)*log(cos(x) - 1)/2 + cos(x) + 1
assert eq.dummy_eq(ans)
```
Though I suspect you know this, I am reiterating for the sake of others: the type and names used in the arg to `dummy_eq` are immaterial (provided that they do not clash with a symbol already in the expression) so the following would also pass:
```python
b1, b2 = symbols('b1 b2')
ans = (-b1 + b2)*log(cos(x) + 1)/2 + (b1 + b2)*log(cos(x) - 1)/2 + cos(x) + 1
assert eq.dummy_eq(ans)
```

Reply:
To me that seems like changing the definition of `dummy_eq`.

For testing what we need is just a function that compares two expressions as equal under a renaming of the Dummy symbols. I wouldn't want to use this version of `dummy_eq` in testing if it gives things like
```julia
In [2]: Symbol('x').dummy_eq(Symbol('y'))                                                                                                      
Out[2]: True
```

Reply:
Hopefully you are not getting that with this branch:

```python
>>> Symbol('x').dummy_eq(Symbol('y'))
False
```

The default behavior of the method is to do as you said: see if Dummy variables in self can be uniquely mapped to symbols in other to give two expressions that are the same. The behavior can be slightly adapted, however, by providing `symbol` in which case the mapping goes from the specified symbols in `other` to `self`.

Reply:
I thought it was this branch but can't reproduce it now. Maybe it was before the force-push...

Okay I see what you mean. I hadn't noticed that `d` was a `Dummy` above. Maybe rename it to `_d`?

Reply:
I do that in the doctest which are more public-facing. In the test suite `d = Dummy()` appears just before the test so I think it's ok.

Reply:
> Maybe it was before the force-push...

I didn't think of that til later -- yes, I corrected that condition and added coverage apparently after you had pulled a version.